casper/src/rust/engine/block_approver_protocol.rs

```diff
@@ -0,0 +1,285 @@
+// See casper/src/main/scala/coop/rchain/casper/engine/BlockApproverProtocol.scala
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use comm::rust::peer_node::PeerNode;
+use comm::rust::rp::rp_conf::RPConf;
+use comm::rust::transport::transport_layer::{Blob, TransportLayer};
+use crypto::rust::hash::blake2b256::Blake2b256;
+use log::{info, warn};
+use models::rust::casper::protocol::casper_message::{
+    ApprovedBlockCandidate, BlockApproval, ProcessedDeploy, ProcessedSystemDeploy, UnapprovedBlock,
+};
+use models::rust::casper::protocol::packet_type_tag::ToPacket;
+use prost::bytes::Bytes;
+use prost::Message;
+
+use crate::rust::errors::CasperError;
+use crate::rust::genesis::contracts::{
+    proof_of_stake::ProofOfStake, validator::Validator, vault::Vault,
+};
+use crate::rust::util::rholang::runtime_manager::RuntimeManager;
+use crate::rust::validator_identity::ValidatorIdentity;
+
+/// Rust port of `coop.rchain.casper.engine.BlockApproverProtocol` from Scala.
+/// The field layout and logic mirror the original as closely as possible.
+#[derive(Clone)]
+pub struct BlockApproverProtocol<T: TransportLayer + Send + Sync + 'static> {
+    // Configuration / static data
+    validator_id: ValidatorIdentity,
+    deploy_timestamp: i64,
+    vaults: Vec<Vault>,
+    bonds: HashMap<crypto::rust::public_key::PublicKey, i64>,
+    bonds_bytes: HashMap<Bytes, i64>, // helper map keyed by raw bytes
+    minimum_bond: i64,
+    maximum_bond: i64,
+    epoch_length: i32,
+    quarantine_length: i32,
+    number_of_active_validators: i32,
+    required_sigs: i32,
+    pos_multi_sig_public_keys: Vec<String>,
+    pos_multi_sig_quorum: i32,
+
+    // Infrastructure
+    transport: Arc<T>,
+    conf: Arc<RPConf>,
+}
+
+impl<T: TransportLayer + Send + Sync + 'static> BlockApproverProtocol<T> {
+    /// Corresponds to Scala `BlockApproverProtocol.of` – constructor with basic validation.
+    pub fn new(
+        validator_id: ValidatorIdentity,
+        deploy_timestamp: i64,
+        vaults: Vec<Vault>,
+        bonds: HashMap<crypto::rust::public_key::PublicKey, i64>,
+        minimum_bond: i64,
+        maximum_bond: i64,
+        epoch_length: i32,
+        quarantine_length: i32,
+        number_of_active_validators: i32,
+        required_sigs: i32,
+        pos_multi_sig_public_keys: Vec<String>,
+        pos_multi_sig_quorum: i32,
+        transport: Arc<T>,
+        conf: Arc<RPConf>,
+    ) -> Result<Self, CasperError> {
+        if bonds.len() <= required_sigs as usize {
+            return Err(CasperError::RuntimeError(
+                "Required sigs must be smaller than the number of bonded validators".to_string(),
+            ));
```

Comment:
Optional: More Explicit Error Message

```suggestion
            return Err(CasperError::RuntimeError(format!(
                "Required sigs ({}) must be smaller than the number of bonded validators ({})",
                required_sigs,
                bonds.len()
            )));
```